query_executor.go

```diff
@@ -50,51 +70,148 @@ func (q *queryExecutor) executeQuery(qry ExecutableQuery) (*Iter, error) {
 			continue
 		}
 
-		iter = q.attemptQuery(qry, conn)
-		// Update host
-		hostResponse.Mark(iter.err)
+		// store the result
+		q.specWG.Add(1)
+		go q.runAndStore(qry, conn, selectedHost, results, nil)
+
+		// fetch from the channel whatever is there
+		res = <-results
 
-		if rt == nil {
-			iter.host = host
-			break
+		if res.retryType == RetryNextHost {
+			continue
 		}
 
-		switch rt.GetRetryType(iter.err) {
-		case Retry:
-			for rt.Attempt(qry) {
-				iter = q.attemptQuery(qry, conn)
-				hostResponse.Mark(iter.err)
-				if iter.err == nil {
-					iter.host = host
-					return iter, nil
-				}
-				if rt.GetRetryType(iter.err) != Retry {
-					break
-				}
-			}
-		case Rethrow:
-			return nil, iter.err
-		case Ignore:
-			return iter, nil
-		case RetryNextHost:
-		default:
+		return res.iter, res.err
+
+	}
+
+	if res.iter == nil {
+		return nil, ErrNoConnections
+	}
+
+	return res.iter, nil
+}
+
+func (q *queryExecutor) executeSpeculativeQuery(qry ExecutableQuery) (*Iter, error) {
+	hostIter := q.policy.Pick(qry)
+	sp := qry.speculativeExecutionPolicy()
+
+	results := make(chan queryResponse)
```

Comment:
buffer this to the number of possible executions so goroutines dont leak

Reply:
Thanks, that's actually a great idea.